openai/client.py

```diff
@@ -0,0 +1,392 @@
+import time
+import typing
+
+import openai
+
+LATEST_AZURE_API_VERSION = "2023-05-15"
+
+
+class AzureTokenAuth:
+    """Authentication using an Azure AD token.
+    """
+
+    def __repr__(self):
+        return f"AzureTokenAuth({type(self._credential)})"
+
+    def __init__(self, *, credential=None):
+        """Create a new AzureTokenAuth instance. If no credential is passed, 
+        it will use ~azure.identity.DefaultAzureCredential
+        """
+        if not credential:
+            try:
+                import azure.identity
+            except ImportError:
+                raise Exception(
+                    "You have to install the azure-identity package in order to use AzureTokenAuth"
+                )
+            credential = azure.identity.DefaultAzureCredential()
+        self._credential = credential
+        self._cached_token = None
+
+    def get_token(self) -> str:
+        if (
+            self._cached_token is None
+            or (self._cached_token.expires_on - time.time()) < 300
+        ):
+            self._cached_token = self._credential.get_token(
+                "https://cognitiveservices.azure.com/.default"
+            )
+
+        return self._cached_token.token
+
+
+class ApiKeyAuth:
+    """Authentication using an API key"""
+
+    def __repr__(self):
+        return f"ApiKeyAuth(api_key=<redacted>)"
+    
+    def __init__(self, key: str = ""):
+        """Create a new ApiKeyAuth instance. If no key is passed, it will use ~openai.api_key"""
+        self.key = key or openai.api_key
+
+    def get_token(self) -> str:
+        """Get the API key"""
+        return self.key
+
+
+Backends = typing.Literal["azure", "openai", ""]
+
+
+class OpenAIClient:
+    def __init__(
+        self,
+        *,
+        api_base: str = "",
+        auth: typing.Union[str, ApiKeyAuth, AzureTokenAuth] = "",
+        api_version: str = "",
+        backend: Backends = "",
+    ):
+        """Create a new OpenAI client.
+           
+           :param: api_base: The base URL for the API. If not specified, based on ~opeanai.api_base
+           :param: auth: The authentication method or key to use. If the string value "azuredefault" is passed, it will use ~azure.identity.DefaultAzureCredential
+           :param: api_version: The API version to use. If not specified, based on ~openai.api_version or ~openai.client.LATEST_AZURE_API_VERSION.
+           :param backend: One of 'azure' or 'openai'. If not specified, inferred from the auth method or ~openai.api_type
```

Comment:
```suggestion
       :keyword str api_base: The base URL for the API. If not specified, based on ~opeanai.api_base
       :keyword auth: The authentication method or key to use. If the string value "azuredefault" is passed, it will use ~azure.identity.DefaultAzureCredential
       :paramtype auth: str or ~openai.client.ApiKeyAuth or ~openai.client.AzureTokenAuth
       :keyword str api_version: The API version to use. If not specified, based on ~openai.api_version or ~openai.client.LATEST_AZURE_API_VERSION.
       :keyword str backend: One of 'azure' or 'openai'. If not specified, inferred from the auth method or ~openai.api_type
```